File: allpathsfromsourceleadtodestination.js
//Objective is to see whether a path exists from a source to a destination node

let n = 3, edges = [[0,1],[0,2]], source = 0, destination = 2


//O(n) solution that does a DFS traversal over the graph to see if a path exists

let graph = new Map()
    
//Build the graph
for (let [v1, v2] of edges) {
    if (!graph.has(v1)) {
        graph.set(v1, new Set())
    }
    graph.get(v1).add(v2)
}

//The destination cannot have any outgoing edges
if (graph.has(destination)) {
    return false
}

let visited = new Set()

//DFS through graph
function dfs(node) {
    //Once we reach a node without an outgoing edge, it must
    //be the destination
    if (!graph.has(node)) {
        return node == destination
    }
    
    for (let neighbor of graph.get(node)) {
        //Check for cycle
        if (visited.has(neighbor)) {
            return false
        }
        visited.add(neighbor)
        
        //Check all neighbors
        if (!dfs(neighbor)) {
            return false
        }
        visited.delete(neighbor)
    }
    
    return true
}
return dfs(source)

File: allpathsfromsourceleadtodestination.java
//Java Solution

class Solution {
    public boolean leadsToDestination(int n, int[][] edges, int source, int destination) {
        Map<Integer, List<Integer>> map = new HashMap<>();
        
        for (int[] edge : edges) {
            if (!map.containsKey(edge[0])) {
                map.put(edge[0], new ArrayList<Integer>());
            }
            map.get(edge[0]).add(edge[1]);
        }
        
        if (map.containsKey(destination)) {
            return false;
        }
        
        Set<Integer> visited = new HashSet<>();
        return dfs(source, visited, map, destination);
    }
    
    public boolean dfs(int node, Set<Integer> visited, Map<Integer, List<Integer>> map, int destination) {
        if (!map.containsKey(node)) {
            return node == destination;
        }
        
        for (int i = 0; i < map.get(node).size(); i++) {
            int neighbor = map.get(node).get(i);
            
            if (visited.contains(neighbor)) {
                return false;
            }
            
            visited.add(neighbor);
            
            if (!dfs(neighbor, visited, map, destination)) {
                return false;
            }
            
            visited.remove(neighbor);
        }
        
        return true;
    }
}